---
title: "single-cell ATAC-seq R Notebook (from output matrices of scATAK) for human brain data GSM4441823"
output: html_notebook
---

# Install necessary software packages
```{r}

#install.packages("Seurat")

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("DropletUtils")

```

# Define cell_selection function to make knee plot and remove empty droplets 
```{r, include=FALSE}

library(dplyr)
library(DropletUtils)
library(Matrix)



cell_selection <- function(dir, mtx, feature, bc, method) {
  data.mtx<-readMM(paste(dir, mtx, sep=""))
  feature<-read.table(paste(dir, feature, sep=""), header=F)
  bc<-read.table(paste(dir, bc, sep=""), header=F)
  set.seed(100)
  
  if (method=="cellranger") {
    colnames(data.mtx)<-gsub('-.*', '', bc$V1)
    rownames(data.mtx)<-paste(feature$V1, ":", feature$V2, "-", feature$V3, sep="")
    br.out <- barcodeRanks(data.mtx)
    e.out <- emptyDrops(data.mtx)
  } else {
    rownames(data.mtx)<-bc$V1
    colnames(data.mtx)<-feature$V1
    br.out <- barcodeRanks(t(data.mtx))
    e.out <- emptyDrops(t(data.mtx))
  }
  

  is.cell <- e.out$FDR <= 1e-5
  sum(is.cell, na.rm=TRUE)
  e.out_sel<-e.out[(!is.na(e.out$FDR) & e.out$FDR<=1e-5),]
  if (method=="cellranger") {
    data.mtx_sel <- as.data.frame(as.matrix(data.mtx[,rownames(e.out_sel)]))
  } else {
    data.mtx_sel <- as.data.frame(as.matrix(data.mtx[rownames(e.out_sel),]))
  }
  
  return(list("bc_rank"=br.out, "mtx_sel"=data.mtx_sel))
}

```


# Download the matrices; Remove empty droplets from scATAK Peak count matrix 
```{r, include=FALSE}

system("wget https://data.caltech.edu/tindfiles/serve/4dc91cdb-f827-4ada-b5c8-d36dfa5bd16f/", ignore.stdout = TRUE)
system("mv index.html mtx_data_hbrain.tar.gz", ignore.stdout = TRUE)
system("tar -xvzf mtx_data_hbrain.tar.gz", ignore.stdout = TRUE)

dir1<-"./mtx_data_hbrain/atac_regions/atac_HIP1/"
mtx<-"atac.mtx"
feature<-"atac.genes.txt"
bc<-"atac.barcodes.txt"
scatak_peak_sel<-cell_selection(dir1, mtx, feature, bc, "scatak")

```


# Create Seurat object
```{r}
library(Seurat)
set.seed(1234)
atac_scatak.mtx<-scatak_peak_sel$mtx_sel
atac_brain_scatak <- CreateSeuratObject(counts = t(atac_scatak.mtx), assay = "scATAK", project = "10x_ATAC")

#total 6244 single cells, variable feature at least detected in more than 10% cells
VariableFeatures(atac_brain_scatak) <- names(which(colSums(atac_scatak.mtx > 0) > 624))
atac_brain_scatak <- RunLSI(atac_brain_scatak, n = 50, scale.max = NULL)
atac_brain_scatak <- RunUMAP(atac_brain_scatak, reduction = "lsi", dims = 1:50)
atac_brain_scatak <- FindNeighbors(object = atac_brain_scatak, reduction = 'lsi', dims = 1:50)
atac_brain_scatak <- FindClusters(object = atac_brain_scatak, verbose = FALSE)
library(ggplot2)
DimPlot(object = atac_brain_scatak, label = TRUE) + NoLegend() + ggtitle("scATAK")

```


# Gene count processing
```{r}

dir2<-"./mtx_data_hbrain/atac_genes/atac_HIP1/"
scatak_gene.mtx<-readMM(paste(dir2, "gene.mtx", sep=""))
feature<-read.table(paste(dir2, "gene.genes.txt", sep=""), header=F)
bc<-read.table(paste(dir2, "gene.barcodes.txt",sep=""), header=F)
rownames(scatak_gene.mtx)<-bc$V1
colnames(scatak_gene.mtx)<-feature$V1
scatak_gene.mtx_sel<-scatak_gene.mtx[rownames(atac_scatak.mtx),]
gene_scatak.merge_sel<-data.frame(scatak_gene.mtx_sel)

```



```{r}

sig_genes<-c("SLC17A7", "GAD2", "MAG", "PDGFRA", "GFAP", "CX3CR1")

atac_brain_scatak <- AddMetaData(atac_brain_scatak, gene_scatak.merge_sel[,sig_genes], sig_genes)
VlnPlot(atac_brain_scatak, features=sig_genes, pt.size=0, ncol=3)
figure3a<-VlnPlot(atac_brain_scatak, features=sig_genes, pt.size=0, ncol=3)

```



```{r}

new.cluster.ids <- c("Oligodendrocytes", "Oligodendrocytes", "Astrocytes", "OPC", "Microglia", "eNeurons", "eNeurons", "eNeurons", "eNeurons", "Astrocytes",  "iNeurons", "eNeurons", "iNeurons")
names(new.cluster.ids) <- levels(atac_brain_scatak)
atac_brain_scatak <- RenameIdents(atac_brain_scatak, new.cluster.ids)
DimPlot(atac_brain_scatak, reduction = "umap", label = TRUE, pt.size = 0.5, label.size=6) + NoLegend()

library(scales)
cell_type_count<-data.frame(table(Idents(atac_brain_scatak)))

colnames(cell_type_count)<-c("Cell_type", "Counts")
cell_type_count$Cell_Percentage<-paste(cell_type_count$Cell_type, "  ", formatC(cell_type_count$Counts/sum(cell_type_count$Counts)*100, format="f", digits=2), "%",  sep="")

figure3b<-ggplot(cell_type_count, aes(x="",y=Counts, fill=Cell_Percentage)) + geom_bar(width = 1, stat = "identity") + coord_polar("y", start=0) + xlab("") + ylab("") + ggtitle(paste("Totol number of cells = ", sum(cell_type_count$Counts))) + theme(plot.title = element_text(hjust = 0.5, size=20), legend.title=element_text(size=14),  legend.text=element_text(size=12))

figure3b + theme(axis.text.x=element_blank())

#figure3b<-DimPlot(atac_brain_scatak, reduction = "umap", label = TRUE, pt.size = 0.5, label.size=3) + NoLegend()

library(cowplot)
png("Figure3_AB.png", width=1200, height=500)
plot_grid(figure3a, figure3b, ncol=2)
dev.off()

```


# Generate barcode-group table
```{r}

bc_group<-data.frame(Barcodes=colnames(atac_brain_scatak), Groups=Idents(atac_brain_scatak))
write.table(bc_group, file="./bc_group.txt", quote=F, row.names=F)
  
```


#Generate Figure3C - circos plot for brain scATAC group tracks and GWAS track

```{r, include=FALSE}

system("wget https://data.caltech.edu/tindfiles/serve/6ecc1ffb-221d-4f59-a698-4828ad77a935/", ignore.stdout = TRUE)
system("mv index.html track_data_hbrain.tar.gz", ignore.stdout = TRUE)
system("tar -xvzf track_data_hbrain.tar.gz", ignore.stdout = TRUE)

library(ComplexHeatmap)
library(circlize)
#generate cell-type-specific ATAC-seq tracks (see data preprocessing tutorial)

# Calculate ATACseq signals in 200Kb windows
system("bash track_data_hbrain/run_bedgraph_bin.sh track_data_hbrain/HIP1_Astrocytes.bedgraph 200000 > track_data_hbrain/HIP1_Astrocytes_bin.bedgraph")
system("bash track_data_hbrain/run_bedgraph_bin.sh track_data_hbrain/HIP1_Microglia.bedgraph 200000 > track_data_hbrain/HIP1_Microglia_bin.bedgraph")
system("bash track_data_hbrain/run_bedgraph_bin.sh track_data_hbrain/HIP1_eNeurons.bedgraph 200000 > track_data_hbrain/HIP1_eNeurons_bin.bedgraph")
system("bash track_data_hbrain/run_bedgraph_bin.sh track_data_hbrain/HIP1_iNeurons.bedgraph 200000 > track_data_hbrain/HIP1_iNeurons_bin.bedgraph")
system("bash track_data_hbrain/run_bedgraph_bin.sh track_data_hbrain/HIP1_Oligodendrocytes.bedgraph 200000 > track_data_hbrain/HIP1_Oligodendrocytes_bin.bedgraph")
system("bash track_data_hbrain/run_bedgraph_bin.sh track_data_hbrain/HIP1_OPC.bedgraph 200000 > track_data_hbrain/HIP1_OPC_bin.bedgraph")

gwas<-read.table("track_data_hbrain/AD_SNP_p_1E-9_hg38_snp151_rev.bed")
gwas$logp<-log10(gwas$V5)*(-1)
gwas_bed<-gwas[,c(1:3,6)]

bed1<-read.table("track_data_hbrain/HIP1_Astrocytes_bin.bedgraph")
bed2<-read.table("track_data_hbrain/HIP1_eNeurons_bin.bedgraph")
bed3<-read.table("track_data_hbrain/HIP1_iNeurons_bin.bedgraph")
bed4<-read.table("track_data_hbrain/HIP1_Microglia_bin.bedgraph")
bed5<-read.table("track_data_hbrain/HIP1_Oligodendrocytes_bin.bedgraph")
bed6<-read.table("track_data_hbrain/HIP1_OPC_bin.bedgraph")

bed1[bed1$V4>mean(bed1$V4)*10, 4] <- mean(bed1$V4)*10
bed2[bed2$V4>mean(bed2$V4)*10, 4] <- mean(bed2$V4)*10
bed3[bed3$V4>mean(bed3$V4)*10, 4] <- mean(bed3$V4)*10
bed4[bed4$V4>mean(bed4$V4)*10, 4] <- mean(bed4$V4)*10
bed5[bed5$V4>mean(bed5$V4)*10, 4] <- mean(bed5$V4)*10
bed6[bed6$V4>mean(bed6$V4)*10, 4] <- mean(bed6$V4)*10


png("circos.png", width=640, height=640)
circos.par("track.height" = 0.07)
circos.initializeWithIdeogram(species = "hg38", chromosome.index = paste0("chr", c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22)))
text(0, 0, "", cex = 1)

circos.genomicTrack(gwas_bed, ylim = c(9, 50), panel.fun = function(region, value, ...) {
  circos.genomicPoints(region, value, cex=0.2, pch=20, col="black", baseline="bottom")})

circos.genomicTrack(bed1, panel.fun = function(region, value, ...) {
  circos.genomicLines(region, value, type="h", col="red", baseline="bottom")})

circos.genomicTrack(bed2, panel.fun = function(region, value, ...) {
  circos.genomicLines(region, value, type="h", col="green", baseline="bottom")})

circos.genomicTrack(bed3, panel.fun = function(region, value, ...) {
  circos.genomicLines(region, value, type="h", col="blue", baseline="bottom")})

circos.genomicTrack(bed4, panel.fun = function(region, value, ...) {
  circos.genomicLines(region, value, type="h", col="cyan", baseline="bottom")})

circos.genomicTrack(bed5, panel.fun = function(region, value, ...) {
  circos.genomicLines(region, value, type="h", col="magenta", baseline="bottom")})

circos.genomicTrack(bed6, panel.fun = function(region, value, ...) {
  circos.genomicLines(region, value, type="h", col="yellow", baseline="bottom")})

lgd_lines = Legend(at = c("AD_GWAS", "Astrocytes", "eNeurons", "iNeurons", "Microglia", "Oligodendrocytes", "OPC"), type = "points", 
     legend_gp = gpar(col = 1:7), title_position = "topleft", title = "Tracks")
draw(lgd_lines)

dev.off()

```


#Generate Figure3D - chromatin accessibility around AD GWAS SNPs (p<1E-9) in different cell types.

```{r}

bin<-2000
gwas$bin<-paste(gwas$V1, ":", as.integer(gwas$V2/bin)*bin, "-", as.integer(gwas$V3/bin + 1)*bin, sep="")

for (i in c("Astrocytes", "Microglia", "eNeurons", "iNeurons", "Oligodendrocytes", "OPC")){
  system(paste("bash track_data_hbrain/run_bedgraph_bin.sh track_data_hbrain/HIP1_", i, ".bedgraph ", bin, " > track_data_hbrain/HIP1_", i, "_bin2k.bedgraph", sep=""))
}

bed_snp_overlap<-NULL  
for (i in c("Astrocytes", "Microglia", "eNeurons", "iNeurons", "Oligodendrocytes", "OPC")){
  bed<-read.table(paste("track_data_hbrain/HIP1_", i, "_bin2k.bedgraph", sep=""))
  bed$bin<-paste(bed$V1, ":", as.integer(bed$V2/bin)*bin, "-", as.integer(bed$V3/bin)*bin, sep="")
  bed_i<-merge(bed, gwas, by="bin")
  bed_i$cell_type<-rep(i, nrow(bed_i))
  bed_snp_overlap<-rbind(bed_snp_overlap, bed_i)
}

gwas_atac<-bed_snp_overlap[bed_snp_overlap$V5>0, c(1,5,9,10,12)]
colnames(gwas_atac)<-c("bin", "ATAC_signal", "SNP", "Pval", "Cell_type")  

gwas_atac$logP<-log10(gwas_atac$Pval)*(-1)

library(tidyverse)
gwas_atac_table<-gwas_atac[,c(5,3,2)] %>%
  spread(key=SNP, value=ATAC_signal)
gwas_atac_table[is.na(gwas_atac_table)]<-0

rownames(gwas_atac_table)<-gwas_atac_table$Cell_type
gwas_atac_table<-gwas_atac_table[,-1]

library(reshape2)
gwas_atac_new<-melt(as.matrix(gwas_atac_table))

colnames(gwas_atac_new)<-c("Cell_type", "SNP", "ATAC_signal")

plot_density<-ggplot(gwas_atac_new, aes(colour=Cell_type, x=log10(ATAC_signal+1))) + geom_density() + scale_color_manual(values=c("red", "green", "blue", "cyan", "magenta", "yellow")) + ggtitle("Chromatin accessibility of AD SNPs") + theme(plot.title = element_text(hjust = 0.5, size=20), legend.title=element_text(size=14),  legend.text=element_text(size=12), text = element_text(size=20))

plot_density
png("gwas_atac_density.png", width=540, height=640)
plot_density
dev.off()

```


# HiC interaction potential analysis
```{r, include=FALSE}

library(dplyr)
library(Seurat)
library(Matrix)
# generate predicted hic interaction matrix (see data preprocessing tutorial)

system("wget https://data.caltech.edu/tindfiles/serve/14e3fed6-f2a7-4950-8e97-431c9b8ce928/", ignore.stdout = TRUE)
system("mv index.html hic_data_hbrain.tar.gz", ignore.stdout = TRUE)
system("tar -xvzf hic_data_hbrain.tar.gz", ignore.stdout = TRUE)
```

# Get cell-type-specific signature chromatin interactions and generate a heatmap (Figure 4A)
```{r}
bc_group<-read.table("./bc_group.txt", header=T)
atac.mtx<-readMM("hic_data_hbrain/HIP1/hic.mtx")
hic<-read.table("hic_data_hbrain/HIP1/hic.genes.txt",header=F)
bc<-read.table("hic_data_hbrain/HIP1/hic.barcodes.txt",header=F)
rownames(atac.mtx)<-bc$V1
colnames(atac.mtx)<-hic$V1
atac_sel.mtx<-atac.mtx[bc_group$Barcodes,]
atac_hic <- CreateSeuratObject(counts = t(atac_sel.mtx), assay = "ATAC_HIP_hic", project = "10x_ATAC")
Idents(atac_hic) <- bc_group$Groups

signature_top5<-NULL

for (i in c("Astrocytes", "eNeurons", "iNeurons", "Microglia", "Oligodendrocytes", "OPC"))
{ 
  cluster_signature <- FindMarkers(object = atac_hic, ident.1 = i, only.pos = TRUE, test.use = "wilcox")
  cluster_signature <- cluster_signature[cluster_signature$p_val_adj<0.05,]
  write.table(cluster_signature, paste("hic_markers_",i,".txt", sep=""), quote=F, sep="\t")
  signature_top5<-append(signature_top5, rownames(cluster_signature)[1:5])
}

my_levels <- c("Astrocytes", "eNeurons", "iNeurons", "Microglia", "Oligodendrocytes", "OPC")
atac_hic@active.ident <- factor(x = atac_hic@active.ident, levels = my_levels)

plot_heatmap<-DoHeatmap(object = subset(atac_hic, downsample=100), features = signature_top5, size = 3, disp.min = 0, disp.max = 2, slot="data") + scale_fill_gradientn(colors = c("blue", "red"))
plot_heatmap

png("hic_heatmap.png", width = 1000, height = 600)
plot_heatmap
dev.off()

```

#Find APOE interactions and generate a UMAP (Figure 4C)
```{r}

interact_sel=c("chr19:44900000-44910000_chr19:44950000-44960000")
APOE=as.matrix(atac_sel.mtx[,interact_sel])

interact_name=c("chr19.44900000.44910000_chr19.44950000.44960000")
atac_brain_scatak <- AddMetaData(atac_brain_scatak, APOE, interact_name)

#plot_feature=FeaturePlot(atac_brain_scatak, reduction="umap", features = interact_sel2, cols=c("lightgrey", "red"))
plot_violin<-VlnPlot(atac_brain_scatak, features=interact_sel2, pt.size=0)
plot_violin
png("./violin_hic.png", width = 500, height = 360)
plot_violin
dev.off()


```



